firedrake/dmplex.pyx

```diff
@@ -1014,3 +964,784 @@ def plex_renumbering(PETSc.DM plex, np.ndarray[PetscInt, ndim=1, mode="c"] reord
+                        0: orientation is same as in (local) plex
+                       +1: orientation is the opposite of that in plex
+                       THIS ARRAY IS UPDATED BY THIS FUNCTION.
+
+    Returns an array of size 'nfacets_shared', which tells for each shared
+    facet which other shared facet needs update, if any, when a shared facet
+    is flipped.
+     * Equal to 'nfacets_shared': no other facet requires update.
+     * Non-negative value: index of shared facet,
+                           which must have the same global orientation.
+     * Negative value 'i': cneg(i) is the index of the shared facet,
+                           which must have opposite global orientation.
+    """
+    cdef:
+        PetscInt nfacets, fStart, fEnd, f
+        PetscInt size
+        PetscInt support[2]
+        PetscInt start_facet, end_facet
+        np.int8_t twist
+        PetscInt i, j
+        np.ndarray[np.int32_t, ndim=1, mode="c"] result
+
+    fStart, fEnd = plex.getHeightStratum(1)
+    nfacets = fEnd - fStart
+
+    result = np.empty(nfacets_shared, dtype=np.int32)
+
+    for f in range(fStart, fEnd):
+        if orientations[f - fStart] < 0:
+            orientations[f - fStart] = 0
+
+            plex_get_restricted_support(plex, cell_ranks, f, &size, support)
+            assert 0 <= size <= 2
+            if size == 0:
+                continue
+
+            end_facet = traverse_cell_string(plex, f, support[0],
+                                             cell_ranks, orientations)
+            if end_facet == -1:
+                # Closed loop
+                if orientations[f - fStart]:
+                    # Moebius strip found
+                    raise RuntimeError("Moebius strip found in the mesh.")
+            else:
+                if size == 1:
+                    # 'f' is at local domain boundary
+                    start_facet = f
+                else:
+                    start_facet = traverse_cell_string(plex, f, support[1],
+                                                       cell_ranks, orientations)
+
+                i = facet2index[start_facet - fStart]
+                j = facet2index[end_facet - fStart]
+                if i >= 0 or j >= 0:
+                    twist = 0
+                    twist ^= get_global_edge_orientation(plex,
+                                                         vertex_numbering,
+                                                         start_facet)
+                    twist ^= orientations[start_facet - fStart]
+                    twist ^= orientations[end_facet - fStart]
+                    twist ^= get_global_edge_orientation(plex,
+                                                         vertex_numbering,
+                                                         end_facet)
+
+                    if i == -1:
+                        result[j] = nfacets_shared
+                    elif j == -1:
+                        result[i] = nfacets_shared
```

Comment:
The other end of the string is local (not shared) so no propagation to remote ranks at the other end is needed.